dictionaries/CallMap.php

```diff
@@ -1706,7 +1706,7 @@
 'cyrus_connect' => ['resource', 'host='=>'string', 'port='=>'string', 'flags='=>'int'],
 'cyrus_query' => ['array', 'connection'=>'resource', 'query'=>'string'],
 'cyrus_unbind' => ['bool', 'connection'=>'resource', 'trigger_name'=>'string'],
-'date' => ['string|false', 'format'=>'string', 'timestamp='=>'int'],
+'date' => ['string', 'format'=>'string', 'timestamp='=>'?int'],
```

Comment:
Callmap.php should contain the signature for the oldest version supported so you should keep |false and not allow nullable. Also the definition in callmap.php must be strictly identical tp tthe 'old' section of the next Callmap (so 8.0 in your case)

Reply:
I think you have this reversed.

Psalm/Internal/Codebase/InternalCallMapHandler loads the current CallMap, and then walks backward from 8.0 through the delta files, removing 'new' entries, and adding 'old' entries until it reaches the version it's analyzing for.

So the main CallMap should have the most recent signature present in the delta files (in this case the 8.0 signature)

Reply:
[date_format](https://www.php.net/manual/en/datetime.format.php) got mapped for php 8, so maybe we can use it as an example.
<details>
<summary>date_format example</summary>

Changelog
![image](https://user-images.githubusercontent.com/52104630/128383018-45e0c40e-77dd-4450-9a01-ff0185882082.png)

new section (php 8.0):
https://github.com/vimeo/psalm/blob/019b35850ac31b3919143ea8e202881e863c925f/dictionaries/CallMap_80_delta.php#L61
old section (php 7.4, I think):
https://github.com/vimeo/psalm/blob/019b35850ac31b3919143ea8e202881e863c925f/dictionaries/CallMap_80_delta.php#L353

And the CallMap.php file seems to have the entry from the old section (NOT the php 8 signature).
```php
'date_format' => ['string|false', 'object'=>'DateTimeInterface', 'format'=>'string'],
```
</details>
This is only one example, but it suggests @orklah is right.

What it doesn't clarify is if we should keep including `|false` in the CallMap.php and the old section.

The docblock for 80_CallMap_delta.php says
```
This contains the information needed to convert the function signatures for php 8.0 to php 7.4 
```
so it seems we would want to put the php 7.4 signature in the old section (ie don't include the `|false`).

The confusion: `|false` seems like the oldest supported version so it should be in CallMap.php. But CallMap.php and the old section of 80_CallMap_delta.php need to match exactly.

> Callmap.php should contain the signature for the oldest version supported so you should keep |false and not allow nullable. Also the definition in callmap.php must be strictly identical tp tthe 'old' section of the next Callmap (so 8.0 in your case)

So, do we include `|false`?

Reply:
I may be confused here, I never get this right :D Honestly, we could use some unit tests for the callmap consistency

Reply:
> I may be confused here, I never get this right :D Honestly, we could use some unit tests for the callmap consistency

[Way ahead of you mate](https://github.com/vimeo/psalm/pull/6247)! :-D

Like I said, I've had a look at the way Psalm reads the CallMap and Delta files. It loads the main callback and then, if the version needed is older than 8.0, starts from the newest delta file and works its way back. This is the logical approach, as it means that stopping support for an old version can be done by simply removing the old delta files. It's also fastest for the newest PHP version, which I imagine is also a plus.

But it looks like a lot of the changes in the 8.0 delta were not applied to the main CallMap, which is why I wrote that test in the first place. I fixed a lot of the errors, but I'm quite sure a lot of the changes to falsey returns are still missing or incorrect.

Unfortunately, there's no comprehensive list of the functions that received that change with the 8.0 release, and the 'false' return type is often absent in the PHP doc even for functions that (still) have it, so the only real way to check it, is to test it.

Reply:
I was confused for a moment by how quick you made these changes after my suggestion after realizing that you posted the PR 3 minutes after my comment and that didn't even left you enough time to write the PR message :D